pkg/policy/selectorcache.go

```diff
@@ -829,78 +353,60 @@ func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identiti
 func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, lbls labels.LabelArray, fqdnSelec api.FQDNSelector) (cachedSelector CachedSelector, added bool) {
 	key := fqdnSelec.String()
 
-	// Lock NameManager before the SelectorCache
+	// Lock NameManager before the SelectorCache. Always.
+	// This is because SelectorCache and NameManager have interleaving locks,
+	// so we must always acquire the NameManager lock first
 	sc.localIdentityNotifier.Lock()
 	defer sc.localIdentityNotifier.Unlock()
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
 
 	// If the selector already exists, use it.
-	sc.mutex.Lock()
-	fqdnSel, exists := sc.selectors[key]
+	idSel, exists := sc.selectors[key]
 	if exists {
-		added := fqdnSel.addUser(user)
-		sc.mutex.Unlock()
-		return fqdnSel, added
+		return idSel, idSel.addUser(user)
 	}
-	sc.mutex.Unlock()
 
-	// Create the new selector. Pulling the identities it selects could
-	// cause allocation of new CIDR identities, so we do this while not
-	// holding the 'sc.mutex'.
-	newFQDNSel := &fqdnSelector{
-		selectorManager: selectorManager{
-			key:              key,
-			users:            make(map[CachedSelectionUser]struct{}),
-			cachedSelections: make(map[identity.NumericIdentity]struct{}),
-			metadataLbls:     lbls,
-		},
+	source := &fqdnSelector{
 		selector: fqdnSelec,
 	}
 
-	// Make the FQDN subsystem aware of this selector and fetch identities
+	// Make the FQDN subsystem aware of this selector and fetch ips
 	// that the FQDN subsystem is aware of.
-	//
-	// If the same 'fqdnSelec' is registered twice here from different
-	// goroutines, we do *NOT* need to unregister the second one because
-	// 'fqdnSelec' is just a struct passed by value. The call below doesn't
-	// retain any references/pointers.
-	//
-	// If this is called twice, one of the results will arbitrarily contain
-	// a real slice of ids, while the other will receive nil. We must fold
-	// them together below.
-	sc.localIdentityNotifier.RegisterForIdentityUpdatesLocked(newFQDNSel.selector)
-	selectors := map[api.FQDNSelector]struct{}{newFQDNSel.selector: {}}
-	_, selectorIPMapping := sc.localIdentityNotifier.MapSelectorsToIPsLocked(selectors)
-
-	// Allocate identities corresponding to the slice of IPs identified as
-	// being selected by this FQDN selector above. This could plausibly
-	// happen twice, once with an empty 'ids' slice and once with the real
-	// 'ids' slice. Either way, they are added to the selector that is
-	// stored in 'sc.selectors[]'.
-	currentlyAllocatedIdentities := sc.allocateIdentityMappings(fqdnSelec, selectorIPMapping)
+	currentIPs := sc.localIdentityNotifier.RegisterForIPUpdatesLocked(source.selector)
```

Comment:
Do you understand the previous "If this was called twice" arguments? Do you know how they now apply with the new logic? I'm not sure the significance of this, but given that the comment is being removed, I'm also not sure whether the complexity goes away because perhaps we're switching to tracking IPs with labels (stateless) vs. identities (stateful).

Reply:
It's even simpler than that -- in the previous version of the code, the NameManager might need to allocate identities during policy calculation. In order for that to succeed, the SelectorCache had to unlock itself first. But, because the SelectorCache was unlocked, it could theoretically get another `AddFQDNPolicy()` of the _same policy_. 

In this version of the code, locks are no longer dropped and it is not possible for a policy to be double-registered. Clean and neat.